### PR TITLE
Add ability to cancel model loading

### PR DESCRIPTION
**Pull request by crasm**
Updated llama_progress_callback so that if it returns false, the model loading is aborted. This is primarily useful for anybody creating UIs that rely on llama.cpp and let users load models on demand.

The test's success condition is getting a nullptr back from llama_load_model_from_file, so it also passes if the model file is missing. Not sure if the test is worth including, to be honest.

### CODE REVIEW

File: tests/test-model-load-cancel.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+#include <cstdlib>
+#include <tuple>
+
+int main(void) {
+    llama_backend_init(false);
+    auto params = llama_model_params{};
+    params.use_mmap = false;
+    params.progress_callback = [](float progress, void * ctx){
+        std::ignore = ctx;
```
Comment (cebtenzzre) [inline tests/test-model-load-cancel.cpp:11]: According to cppreference, "the behavior of std::ignore outside of std::tie is not formally specified."

File: llama.cpp [hunk 2]
```diff
@@ -3712,16 +3722,18 @@ static bool llama_model_load(const std::string & fname, llama_model & model, con
             return true;
```
Comment (cebtenzzre) [inline llama.cpp:3722]: This function is returning true, 0, -1, and -2?

**Comment by crasm** [comment]
@cebtenzzre This should be ready. Can you please take another look? Thank you!

**Review by ggerganov — APPROVED** [review]
Merge after @cebtenzzre approves

File: llama.cpp [hunk 3]
```diff
@@ -9017,11 +9029,18 @@ struct llama_model * llama_load_model_from_file(
...
 
-    if (!llama_model_load(path_model, *model, params)) {
-        LLAMA_LOG_ERROR("%s: failed to load model\n", __func__);
+    int status = llama_model_load(path_model, *model, params);
+    GGML_ASSERT(status <= 0);
+    if (status < 0) {
+        if (status == -1) {
+            LLAMA_LOG_ERROR("%s: failed to load model\n", __func__);
+        } else if (status == -2) {
+            LLAMA_LOG_INFO("%s, cancelled model load\n", __func__);
```
Comment (cebtenzzre) [inline llama.cpp:9042]:
```suggestion
            LLAMA_LOG_INFO("%s: cancelled model load\n", __func__);
```
btw should we be using British English in strings or American English?

Comment (ggerganov) [reply llama.cpp:9042]: No preference, cancelled is fine

**Comment by crasm** [comment]
Sorry, had an idea to check for the model file exists. If this fails CI I can remove it

File: tests/test-model-load-cancel.cpp
Comment (cebtenzzre) [inline tests/test-model-load-cancel.cpp:1]: This test should be added to the Makefile, and we probably *should* find a way to make it pass with a warning if the model isn't found.

Comment (crasm) [reply tests/test-model-load-cancel.cpp:1]: In that case, how about I create a mock/stub gguf that is just metadata with a few fake tensors, and commit that to the repo? I can try to figure that out if you like the idea.

Comment (ggerganov) [reply tests/test-model-load-cancel.cpp:1]: I think it is easier to add it to https://github.com/ggerganov/llama.cpp/tree/master/ci
In some of the tests, we already download a model and it can be used to run this test on it.

Comment (crasm) [reply tests/test-model-load-cancel.cpp:1]: Hmm. All of the CI parts that use models are in `gg_run_open_llama_{3,7}b_v2`. But before those are run, `gg_run_ctest_{debug,release}` are run. None of the existing ctests require models, and downloading models (and running convert.py on them) before running the other tests would probably be a mistake.

If we want to have ctests that run with models rather than mocking out the gguf for this one test, it'll require creating a separate tests directory or subdirectory (and in the cmake/make), and a separate CI run for those tests that happen after `gg_run_open_llama_{3.7}b_v2`.

So it's not trivial to add this test to the CI. However, it might still be worth it, so that it's easy in the future to add CI tests that need the models.

Comment (ggerganov) [reply tests/test-model-load-cancel.cpp:1]: Ah, yes. If we want to go the "mock gguf" way, you can take a look at the `examples/gguf` example - it has functionality to write a gguf file. I guess we can update the `test-model-load-cancel` to write a small "llama" model and then load it with cancellation. But it can get tedious because you would need to populate the entire model - i.e. all required KV meta + all required tensors. The benefit of this would be that we don't have to commit the mock file - it will be generated by the test on-the-fly.

Another approach which seems easier is to add [ctest label](https://cmake.org/cmake/help/latest/prop_test/LABELS.html) to the tests that require a model (for example, label `model`) and then run a `ctest -L model` command inside `gg_run_open_llama_{3,7}b_v2` where we already have model data. The existing `ctest` commands will be changed to run `ctest -LE model` so that they exclude the tests that depend on model data

**Comment by crasm** [comment]
The ctest labels are working great for cmake. However, it's failing on `make test` in `build.yml` and `code-coverage.yml` because there's no model file on the github workflow.

* Should I find a way to selectively run this test in make if it's local vs in the CI/CD pipeline?
* Change the test name and do some globbing? (idk if possible in make)
* Create a new target separate from `make test`? So we'd have `BUILD_TARGETS`, `TEST_TARGETS`, and `TEST_WITH_MODEL_TARGETS`
* ...?

I could also use some feedback on the direction I'm taking the `run.sh` script. I would like to make it more robust, cut down on the repetitive logic duplicated among 3B/7B/debug/release, and have a clean shellcheck—not sure I can do all that in this PR, however.

What state should I be aiming for to get this merged?

File: requirements.txt [hunk 4]
```diff
@@ -1,4 +1,6 @@
 numpy==1.24.4
 sentencepiece==0.1.98
+torch>=2.0.0
```
Comment (ggerganov) [inline requirements.txt:3]: I don't think `torch` is needed

**Review by ggerganov — APPROVED** [review]
The PR is already quite useful, so we should merge it.

Let's for now disable the test target in the Makefile and figure out what is the best way to enable it later - I'm not sure as well.

I like the changes in `ci/run.sh`. The only constraint for this script is to not exceed some runtime - currently the limit is 30 minutes: 

https://github.com/ggml-org/ci/blob/master/env.sh#L113

We can increase it if necessary, but only as a last resort.

Aside from refactoring and shellcheck, the script can be improved with more simple tests that rely on model data and are not suitable for Github Actions - i.e. require GPU for example. Any improvements in this direction would be great.

File: requirements.txt (hunk 4 above)
Comment (crasm) [reply requirements.txt:3]: It otherwise fails with a clean venv, so I needed to add it. I see the note in `convert.py` about dropping it as a dependency, but it is imported in a number of python files.

```
% find . -name '*.py' -not -path './tmp/*' -print0 | xargs -0 -- grep -nFe 'import torch'
./convert-lora-to-ggml.py:12:import torch
./convert-persimmon-to-gguf.py:1:import torch
./convert-hf-to-gguf.py:16:import torch
./examples/llava/llava-surgery.py:4:import torch
./examples/llava/convert-image-encoder-to-gguf.py:5:import torch
```

`./convert-lora-to-ggml.py` is the one that fails CI. I don't see any other python files being called in `run.sh`.

Comment (slaren) [reply requirements.txt:3]: Generally we have used `requirements.txt` for `convert.py` only. We also have a different `requirements-hf-to-gguf.txt` for `convert-hf-to-gguf.py`. I am not sure what changed, but `convert-lora-to-ggml.py` doesn't fail currently, but it does require pytorch.

Comment (crasm) [reply requirements.txt:3]: To clarify, `./convert-lora-to-ggml.py` is called in `run.sh`, and `run.sh` reliably fails with an error if torch is not included in the script's venv.

If it is not failing for you, you must already have pytorch installed in your user pip packages? Before my changes, `run.sh` did not use a virtual environment.

Instead of adding pytorch to `requirements.txt`, I can alternatively do a `pip install torch` in `run.sh` if that's preferable.

**Comment by crasm** [comment]
The new organization for `requirements.txt` will force versions across top-level scripts to remain compatible. This way, including torch as a dependency in both `convert-lora-to-ggml.py` and `convert-hf-to-gguf.py` won't end up with weird mismatches where they need different versions of pytorch.

If there are no objections to the new `requirements.txt` structure or any other notes, I will merge this PR tomorrow.

Btw I am getting test failure locally after updating from master. I don't think it's anything I did because it's unrelated to model loading.

<details>

```
[MUL] NMSE = 0.000000 FAIL
  708/709 tests passed
  Backend Metal: FAIL

1/2 backends passed
FAIL

      Start 19: test-rope
19/19 Test #19: test-rope ...........................   Passed    0.08 sec

95% tests passed, 1 tests failed out of 19

Label Time Summary:
main    =  20.27 sec*proc (19 tests)

Total Test time (real) =  20.28 sec

The following tests FAILED:
         18 - test-backend-ops (Failed)
```

</details>

**Comment by crasm** [comment]
Failing test from re-running `run.sh` without manually cleaning out the artifacts. It shouldn't be necessary to manually clean things up on re-runs, so I'll look into that as I begin revamping the rest of the script.

@cebtenzzre this is ready to go, pending any objections

File: convert_requirements.txt
Comment (ggerganov) [inline convert_requirements.txt:1]: Let's rename all `convert-*_requirements.txt` to `requirements-convert-*.txt`

Note to use dash instead of underscore

**Comment by nittaya111** [comment]
😎 what? My family 😎

เมื่อ พ. 20 ธ.ค. 2566 เวลา 15:40 Georgi Gerganov ***@***.***>
เขียนว่า:

**Comment by LoganDark** [comment]
https://github.com/ggerganov/llama.cpp/blob/799fc2268989482054944c902874cca76337580f/ggml-cuda.cu#L9094

now needs to check that `tensor->extra` is not NULL

image

```diff
-    if (!tensor || (tensor->backend != GGML_BACKEND_GPU && tensor->backend != GGML_BACKEND_GPU_SPLIT) ) {
+    if (!tensor || !tensor->extra || (tensor->backend != GGML_BACKEND_GPU && tensor->backend != GGML_BACKEND_GPU_SPLIT) ) {
```

fixes the access violation

File: llama.cpp [hunk 5]
```diff
@@ -3722,16 +3729,23 @@ static void llm_load_tensors(
         model.tensors_by_name.emplace_back(ggml_get_name(cur), cur);
     }
 
-    ml.load_all_data(ctx, progress_callback, progress_callback_user_data, buf_mmap, use_mlock ? &model.mlock_mmap : NULL);
+    ok = ok && ml.load_all_data(ctx, progress_callback, progress_callback_user_data, buf_mmap, use_mlock ? &model.mlock_mmap : NULL);
+    if (progress_callback) {
+        // Even though the model is done loading, we still honor
+        // cancellation since we need to free allocations.
+        ok = ok && progress_callback(1.0f, progress_callback_user_data);
```
Comment (slaren) [inline llama.cpp:3736]: The last `progress_callback` call is done in `load_all_data` now.

Comment (crasm) [reply llama.cpp]: I missed that. However, that's slightly different behavior. I wanted to make sure that if the loading is aborted even at `1.0f` that it still aborts.

File: llama.cpp [hunk 6]
```diff
@@ -3729,19 +3729,16 @@ static bool llm_load_tensors(
...
-    ok = ok && ml.load_all_data(ctx, progress_callback, progress_callback_user_data, buf_mmap, use_mlock ? &model.mlock_mmap : NULL);
-    if (progress_callback) {
-        // Even though the model is done loading, we still honor
-        // cancellation since we need to free allocations.
-        ok = ok && progress_callback(1.0f, progress_callback_user_data);
+    if (!ml.load_all_data(ctx, progress_callback, progress_callback_user_data, buf_mmap, use_mlock ? &model.mlock_mmap : NULL)) {
+        return false;
     }
 
     model.mapping = std::move(ml.mapping);
```
Comment (crasm) [inline llama.cpp:3736]: @slaren do you know if this line will be a problem? Since it doesn't get run if the above returns early

Comment (LoganDark) [reply llama.cpp:3736]: > @slaren do you know if this line will be a problem? Since it doesn't get run if the above returns early

The progress callback should only be called if it loaded successfully, I think. Would be weird to run it with 1.0 if the model load actually failed

**Comment by crasm** [comment]
Would greatly appreciate this is ok'd for merge sooner rather than later. I am not that confident I can fix merge conflicts without missing things

File: llama.cpp (hunk 6 above)
Comment (slaren) [reply llama.cpp:3736]: Skipping the `mapping` move should be fine.

Comment (crasm) [reply llama.cpp:3736]: The case I'm trying to **avoid** is:
1. The model has fully loaded
2. The client has received some information that the model load should be cancelled (e.g. on another thread)
3. llama.cpp calls the progress_callback with `1.0f`
4. Client returns false (independent from the `1.0f` value itself)
5. llama.cpp ignores the callback return value, and returns a valid `llama_model *`
6. Now the client has to free the model, even though it did not expect one to be returned

**Comment by slaren** [comment]
I tried running the CI locally, but the `test-model-load-cancel` fails due to not finding the model. Is it supposed to work in the CI with the open llama models?

```
    Start 20: test-model-load-cancel
1/1 Test #20: test-model-load-cancel ...........***Failed    0.02 sec
no model found
```

**Comment by crasm** [comment]
> I tried running the CI locally, but the `test-model-load-cancel` fails due to not finding the model. Is it supposed to work in the CI with the open llama models?

It's supposed to. The test itself looks for the Q2_k quant for 3B and 7B in the models-mnt directory.

I'm assuming your `models-mnt/open-llama/7B-v2` directory exists?

I'll try to debug, but it'll be a little tricky because I don't have an Nvidia device operational at the moment (so the compile fails).

**Comment by slaren** [comment]
Yes, but the current directory doesn't seem to be what the test expects.

```
+ gg_run_ctest_with_model
+ cd /home/diego/code/llama.cpp
+ tee /home/diego/code/llama.cpp/tmp/results/ctest_with_model.log
+ cd build-ci-release
+ set -e
+ tee -a /home/diego/code/llama.cpp/tmp/results/ctest_with_model-ctest_with_model.log
+ ctest --output-on-failure -L model
Test project /home/diego/code/llama.cpp/build-ci-release
    Start 20: test-model-load-cancel
1/1 Test #20: test-model-load-cancel ...........***Failed    0.02 sec
```

The models are in `/home/diego/code/llama.cpp/tmp/mnt/models/open-llama/7B-v2/`

**Comment by slaren** [comment]
This path works: `"../../models-mnt/open-llama/3B-v2/ggml-model-q2_k.gguf"`.

File: tests/test-model-load-cancel.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+int main(void) {
+    const char * models_to_try[] = {
+        // Same default as example/main for local use
+        "./models/7B/ggml-model-f16.gguf",
+        // Models for ./ci/run.sh
+        "./models-mnt/open-llama/3B-v2/ggml-model-q2_k.gguf",
+        "./models-mnt/open-llama/7B-v2/ggml-model-q2_k.gguf",
+    };
+
+    const char * chosen_model;
```
Comment (slaren) [inline tests/test-model-load-cancel.cpp:16]:
```suggestion
    const char * chosen_model = nullptr;
```

**Comment by crasm** [comment]
Hmm, I need to go over the CI script again. I seem to be unable to get the test to fail at all, even when I sabotage it—though it fails and succeeds as expected when run manually. May be related to the path issue.

**Comment by crasm** [comment]
I'm going to split this PR up. The precipitating feature (cancelling model load) is nearly trivial and can be individually reviewed and merged.

It's the consequences of integrating the test that are overcomplicating this.

This PR will be updated to contain only the changes to `llama.h` and `llama.cpp`, without the test.

The other PRs will be:
* Requirements.txt checking and updates
* Revamp of `ci/run.sh` and model load test
